### ISSUE TITLE
Network change while running results in broken connection

### ISSUE THREAD

**Issue by swedneck** [hyprspace/hyprspace#1]
The interface remains up on both ends (only tested with desktop and laptop) with no change, but any attempts to ping either device fail.

**Comment by alecbcs** [comment]
Interesting! Is this with a full network change (moving from one public IP address to another) or switching internal IP addresses within a local network (for example going from WiFi to an Ethernet connection)? I haven't had the opportunity to test moving between public IP addresses yet (a.k.a I haven't taken my laptop down to a coffee shop) . But switching between WiFi and an Ethernet connection on my laptop I'm seeing the connection being reconnected to my other machine after about 30s of packets being lost. If you wait a minute or two is the connection eventually reconnected or nothing at all?

**Comment by alecbcs** [comment]
> It does take a painful amount of time though

Hmm let's see if we can't work on making this a bit quicker! It might be a limitation of the Libp2p underlying network stack but I know the guys over there were looking to make AutoNAT (the system for finding and connecting to devices behind NATs/Firewalls) quicker and more robust.

> Also, is there any way to see logs for the interfaces?

Not at the moment but this is a great idea! Were you thinking the interfaces should write their logs to a file? If so, how much information would you like? Peer connects/disconnects?

**Comment by swedneck** [comment]
Made a separate issue for logging.

**Comment by alecbcs** [comment]
I'm not sure this is a complete fix but reconnecting should be much improved after #37 is merged in.

**Comment by alecbcs** [comment]
@mrbluecoat How are your connectivity/reconnection times?

**Comment by mrbluecoat** [comment]
@alecbcs Not great. I just tried a connection test and it's going on 11 minutes and counting without a connection.

```
# systemctl status hyprspace@utun1 | tail
      Tasks: 12 (limit: 28713)
     Memory: 80.6M
     CGroup: /system.slice/system-hyprspace.slice/hyprspace@utun1.service
             └─7238 /usr/local/bin/hyprspace up utun1 --foreground

Jan 28 04:08:37 systemd[1]: Started hyprspace daemon for utun1.
Jan 28 04:08:37 hyprspace[7238]: [+] Creating TUN Device
Jan 28 04:08:38 hyprspace[7238]: [+] Creating LibP2P Node
Jan 28 04:08:38 hyprspace[7238]: [+] Setting Up Node Discovery via DHT
Jan 28 04:08:38 hyprspace[7238]: [+] Network Setup Complete...Waiting on Node Discovery
```

Did the recent crypto crash perhaps have an effect on the DHT network?

**Comment by mrbluecoat** [comment]
P.S. I also tried with [EdgeVPN](https://github.com/mudler/edgevpn) and got a similar result so I think it's DHT

```
# edgevpn --address 10.1.0.12/24
 INFO           edgevpn  Copyright (C) 2021-2022 Ettore Di Giacinto
       This program comes with ABSOLUTELY NO WARRANTY.
       This is free software, and you are welcome to redistribute it
       under certain conditions.
 INFO  Version: v0.8.5 commit: 605278afad3e4aa5cc67842097c9ce68b555661e
 INFO   Starting EdgeVPN network
 INFO   Node ID: 12D3KooWCRikZ1wYU....
 INFO   Node Addresses: [/ip4/192.168.9.190/tcp/41223 /ip4/127.0.0.1/tcp/41223 /ip6/::1/tcp/37571]
 INFO   Bootstrapping DHT
```

**Comment by alecbcs** [comment]
> @alecbcs Not great. I just tried a connection test and it's going on 11 minutes and counting without a connection.
> 
> ```
> # systemctl status hyprspace@utun1 | tail
>       Tasks: 12 (limit: 28713)
>      Memory: 80.6M
>      CGroup: /system.slice/system-hyprspace.slice/hyprspace@utun1.service
>              └─7238 /usr/local/bin/hyprspace up utun1 --foreground
> 
> Jan 28 04:08:37 systemd[1]: Started hyprspace daemon for utun1.
> Jan 28 04:08:37 hyprspace[7238]: [+] Creating TUN Device
> Jan 28 04:08:38 hyprspace[7238]: [+] Creating LibP2P Node
> Jan 28 04:08:38 hyprspace[7238]: [+] Setting Up Node Discovery via DHT
> Jan 28 04:08:38 hyprspace[7238]: [+] Network Setup Complete...Waiting on Node Discovery
> ```
> 
> Did the recent crypto crash perhaps have an effect on the DHT network?

I think this might just be that libp2p takes a while to start announcing a new public IP address. I just released v0.2.0, let me know if that changes anything otherwise it's probably worth implementing some logic to check for a new network and manually rebuild the peers. If you restart the interface does it connect quickly?

**Comment by mrbluecoat** [comment]
Version 0.2.1 connected quickly on my first attempt. I'll keep monitoring, but it looks promising. :star:

**Comment by mrbluecoat** [comment]
FYI, my home network is running 180 Mbps (per fast.com) but when running iperf3 between my home client and a cloud server with gigabit speed network via hyprspace, I'm only getting ~10 Mbps

```
$ iperf3 -c 10.2.3.4
Connecting to host 10.2.3.4, port 5201
[  5] local 10.2.3.3 port 53812 connected to 10.2.3.4 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  2.21 MBytes  18.6 Mbits/sec    0    204 KBytes       
[  5]   1.00-2.00   sec  1.35 MBytes  11.3 Mbits/sec    0    258 KBytes       
[  5]   2.00-3.00   sec  1.04 MBytes  8.74 Mbits/sec    0    313 KBytes       
[  5]   3.00-4.00   sec  1.17 MBytes  9.77 Mbits/sec    0    367 KBytes       
[  5]   4.00-5.00   sec  1.35 MBytes  11.3 Mbits/sec    0    422 KBytes       
[  5]   5.00-6.00   sec  1.04 MBytes  8.74 Mbits/sec    0    477 KBytes       
[  5]   6.00-7.00   sec  1.17 MBytes  9.77 Mbits/sec    0    532 KBytes       
[  5]   7.00-8.00   sec  1.23 MBytes  10.3 Mbits/sec    0    585 KBytes       
[  5]   8.00-9.00   sec  1.41 MBytes  11.8 Mbits/sec    0    640 KBytes       
[  5]   9.00-10.00  sec   753 KBytes  6.17 Mbits/sec    0    695 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  12.7 MBytes  10.7 Mbits/sec    0             sender
[  5]   0.00-10.63  sec  12.1 MBytes  9.55 Mbits/sec                  receiver

iperf Done.
```

**Comment by alecbcs** [comment]
@mrbluecoat from what I believe, [fast.com](https://fast.com) only shows you your internet download speed, not your upload speed which is what an iperf test through hyprspace is going to be dependent on. What is your internet's upload speed through [speedtest.net](https://speedtest.net)?

**Comment by mrbluecoat** [comment]
Good point, upload speed was 10.28 Mbps. What Linux-based bandwidth test do you recommend for hyprspace?

**Comment by mrbluecoat** [comment]
I performed a more reliable test between two cloud servers in the same data center:

```
$ iperf3 -c $PUBLICIP
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   484 MBytes  4.06 Gbits/sec    0   3.05 MBytes       
[  5]   1.00-2.00   sec   490 MBytes  4.11 Gbits/sec  158   2.45 MBytes       
[  5]   2.00-3.00   sec   496 MBytes  4.16 Gbits/sec    0   3.03 MBytes       
[  5]   3.00-4.00   sec   496 MBytes  4.16 Gbits/sec    0   3.03 MBytes       
[  5]   4.00-5.00   sec   496 MBytes  4.16 Gbits/sec    0   3.03 MBytes       
[  5]   5.00-6.00   sec   494 MBytes  4.14 Gbits/sec  142   2.15 MBytes       
[  5]   6.00-7.00   sec   495 MBytes  4.15 Gbits/sec  157   2.35 MBytes       
[  5]   7.00-8.00   sec   496 MBytes  4.16 Gbits/sec    0   3.02 MBytes       
[  5]   8.00-9.00   sec   496 MBytes  4.16 Gbits/sec    0   3.03 MBytes       
[  5]   9.00-10.00  sec   496 MBytes  4.16 Gbits/sec    0   3.03 MBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  4.82 GBytes  4.14 Gbits/sec  457             sender
[  5]   0.00-10.01  sec  4.82 GBytes  4.14 Gbits/sec                  receiver

iperf Done.
```

```
$ iperf3 -c $HYPRSPACEIP
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  41.6 MBytes   349 Mbits/sec    9    566 KBytes       
[  5]   1.00-2.00   sec  46.6 MBytes   391 Mbits/sec    0    633 KBytes       
[  5]   2.00-3.00   sec  48.1 MBytes   404 Mbits/sec    0    684 KBytes       
[  5]   3.00-4.00   sec  50.0 MBytes   420 Mbits/sec    4    509 KBytes       
[  5]   4.00-5.00   sec  42.2 MBytes   354 Mbits/sec    0    568 KBytes       
[  5]   5.00-6.00   sec  48.1 MBytes   403 Mbits/sec    0    627 KBytes       
[  5]   6.00-7.00   sec  46.4 MBytes   389 Mbits/sec    0    676 KBytes       
[  5]   7.00-8.00   sec  52.9 MBytes   443 Mbits/sec   10    517 KBytes       
[  5]   8.00-9.00   sec  47.5 MBytes   398 Mbits/sec    0    592 KBytes       
[  5]   9.00-10.00  sec  51.6 MBytes   433 Mbits/sec    0    647 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   475 MBytes   398 Mbits/sec   23             sender
[  5]   0.00-10.01  sec   474 MBytes   397 Mbits/sec                  receiver

iperf Done.
```

**Comment by alecbcs** [comment]
Nice! That's still lower performance than I'm seeing between two of my systems on a local 1G connection.

```
Connecting to host 10.1.1.2, port 8080
[  5] local 10.1.1.1 port 33302 connected to 10.1.1.2 port 8080
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   103 MBytes   861 Mbits/sec    0    382 KBytes       
[  5]   1.00-2.00   sec  98.1 MBytes   823 Mbits/sec    0    382 KBytes       
[  5]   2.00-3.00   sec  96.9 MBytes   813 Mbits/sec    0    385 KBytes       
[  5]   3.00-4.00   sec  93.6 MBytes   785 Mbits/sec    0    449 KBytes       
[  5]   4.00-5.00   sec  93.8 MBytes   786 Mbits/sec    0    382 KBytes       
[  5]   5.00-6.00   sec  98.7 MBytes   828 Mbits/sec    0    385 KBytes       
[  5]   6.00-7.00   sec  95.7 MBytes   803 Mbits/sec    0    390 KBytes       
[  5]   7.00-8.00   sec  98.0 MBytes   823 Mbits/sec    0    385 KBytes       
[  5]   8.00-9.00   sec  97.7 MBytes   820 Mbits/sec    0    382 KBytes       
[  5]   9.00-10.00  sec  93.2 MBytes   781 Mbits/sec    0    460 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   968 MBytes   812 Mbits/sec    0             sender
[  5]   0.00-10.01  sec   966 MBytes   810 Mbits/sec                  receiver

iperf Done.
```

I'm guessing that your results could possibly be lower because of two things. 1) Are your VMs of a shared CPU or dedicated CPU class? Many cloud providers over schedule shared nodes and so you might be hitting up against the CPU rate limit of your instance. 2) If you're getting a QUIC buffer warning at the start of Hyprspace have you increased your QUIC buffer size by running,
```
sudo sysctl -w net.core.rmem_max=2500000
```

**Comment by mrbluecoat** [comment]
QUIC buffer was set. Using virtual CPUs with the Oracle Cloud arm instances.
